### DISCUSSION TITLE
Immutable `Clean`

### DISCUSSION THREAD

**Discussion by crishoj** [sinclairzx81/typebox#1073] [Q&A] [answered]
The [documentation](https://github.com/sinclairzx81/typebox?tab=readme-ov-file#values-clean) rightfully notes that `Clean` is a mutable operation:

> Use Clean to remove excess properties from a value. This function does not check the value and returns an unknown type. You should Check the result before use. Clean is a mutable operation. To avoid mutation, Clone the value first.

What is the idiomatic way to clean values in a non-destructive way?

Would it be `Value.Clean(structuredClone(...))`, or is there a better option?

**Comment by sinclairzx81** [comment, marked as answer]
@crishoj Hi,

You can use either `structuredClone(...)` or `Value.Clone(...)` prior to calling `Value.Clean(T, ...)`. There isn't really an established idiom for this (just Clone the value before calling mutable Value functions). However, if you're interested, TypeBox internally handles value pipeline operations in the following way (for it's Parse function)

https://github.com/sinclairzx81/typebox/blob/master/src/value/parse/parse.ts#L43-L57

```typescript
const ParseReducer: ReducerFunction[] = [
  (_schema, _references, value) => Clone(value),
  (schema, references, value) => Default(schema, references, value),
  (schema, references, value) => Clean(schema, references, value),
  (schema, references, value) => Convert(schema, references, value),
  (schema, references, value) => { Assert(schema, references, value); return value },
  (schema, references, value) => (HasTransform(schema, references) ? TransformDecode(schema, references, value) : value),
]
// ------------------------------------------------------------------
// ParseValue
// ------------------------------------------------------------------
function ParseValue<T extends TSchema, R = StaticDecode<T>>(schema: T, references: TSchema[], value: unknown): R {
  return ParseReducer.reduce((value, reducer) => reducer(schema, references, value), value) as R
}
```

You can use this as a basis for custom value processing.

Hope this helps!
S
